overlord/devicestate/handlers_systems.go

```diff
@@ -141,42 +143,64 @@ func (m *DeviceManager) doCreateRecoverySystem(t *state.Task, _ *tomb.Tomb) (err
 	if err != nil {
 		return fmt.Errorf("internal error: cannot obtain recovery system setup information")
 	}
+
+	if len(setup.LocalSnaps) != len(setup.LocalSnapPaths) {
+		return fmt.Errorf("internal error: number of side infos and paths must match")
+	}
+
 	label := setup.Label
 	systemDirectory := setup.Directory
 
 	// get all infos
-	infoGetter := func(name string) (info *snap.Info, present bool, err error) {
-		// snaps are either being fetched or present in the system
-
-		if isRemodel {
-			// in a remodel scenario, the snaps may need to be
-			// fetched and thus their content can be different from
-			// what we have in already installed snaps, so we should
-			// first check the download tasks before consulting
-			// snapstate
-			logger.Debugf("requested info for snap %q being installed during remodel", name)
-			for _, tskID := range setup.SnapSetupTasks {
-				taskWithSnapSetup := st.Task(tskID)
-				snapsup, err := snapstate.TaskSnapSetup(taskWithSnapSetup)
-				if err != nil {
-					return nil, false, err
-				}
-				if snapsup.SnapName() != name {
-					continue
-				}
-				// by the time this task runs, the file has already been
-				// downloaded and validated
-				snapFile, err := snapfile.Open(snapsup.MountFile())
-				if err != nil {
-					return nil, false, err
-				}
-				info, err = snap.ReadInfoFromSnapFile(snapFile, snapsup.SideInfo)
-				if err != nil {
-					return nil, false, err
-				}
-
-				return info, true, nil
+	infoGetter := func(name string) (info *snap.Info, path string, present bool, err error) {
+		// snaps will come from one of these places:
+		//   * passed into the task via a list of side infos (these would have
+		//     come from a user posting snaps via the API)
+		//   * have just been downloaded by a task in setup.SnapSetupTasks
+		//   * already installed on the system
+
+		for i, si := range setup.LocalSnaps {
+			if si.RealName != name {
+				continue
+			}
+
+			path := setup.LocalSnapPaths[i]
+
+			// TODO: need to verify the snaps at some point?
+			info, _, err := backend.OpenSnapFile(path, si)
```

Comment:
instead of importing backend it seems this is similar to the code below using snapfile.Open etc?

Reply:
Will do, easy fix. Just for my knowledge, could you explain the hesitation to import `backend`? `backend.OpenSnapFile` does exactly what happens below with `snapfile.Open` and `snap.ReadInfoFromSnapFile`.

Reply:
snapstate/backend is meant to have an API defined by the needs of snapstate. If you look around nothing (except maybe tests?) imports it but snapstate. snapd started before golang had internal packages support, in principle we could make it internal/backend and similar for other packages like it under the state manager packages

Reply:
Makes sense, thanks for the explanation!